Update in-place upgrade design

abhinavmpandey08 opened the pull request:
*Description of changes:*
This PR aims to update the in-place upgrade design doc to reflect the changes in the current design.
Here are some of the main changes in the designs that we made and their justification:

- Relying on annotations on `KCP` and `MD` objects instead of CAPI runtime extensions to trigger in-place upgrades
  - This was changed because the runtime extensions needed for this feature aren't defined in upstream CAPI yet but it will be added when in-place is officially supported upstream. Until then we would have to maintain a patch in our CAPI to add these runtime extensions which may or may not match what upstream CAPI would have. 
    Therefore the initial investment in adding runtime extensions doesn't seem justified considering we can achieve similar functionality by using annotations, which is a 2 way door decision. Once the runtime extension support for in-place upgrade is officially added in upstream CAPI, we can migrate to it without a lot of throwaway code.

- Another decision that was changed was to remove unnecessary fields from the `ControlPlaneUpgrade`, `MachineDeploymentUpgrade` and the `NodeUpgrade` APIs.
  - This was done because we can gather all the information that we need from the KCP/MD ref including the machines that need upgrade, the new Kubernetes version that the machines need to be upgraded to, and versions of other components that need to be upgraded.

- Lastly, the other major change was `NodeUpgrade` controller deciding what commands to execute on a node based on the node type (cp vs worker) instead of being passed a list of commands.
  - This was done because most of the commands were generic between control plane node and worker nodes. The only exception was the `kubeadm upgrade`. So to account for this, the `NodeUpgrade` object will accept a `NodeType` field to signify whether the node to be upgraded should be a cp or a worker. Then the `ControlPlaneUpgrade` and the `MachineDeploymentUpgrade` controller can just pass the `NodeType` down without having to worry about what the exact implementation of what upgrading a control plane node or a worker node would entail, as those details are abstracted away.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

<!-- If this is a security issue, please do not discuss on GitHub. Please report any suspected or confirmed security issues to AWS Security https://aws.amazon.com/security/vulnerability-reporting/ -->

vivek-koppuru commented on designs/in-place-upgrades-eksa.md:
```diff
@@ -79,14 +83,51 @@ There are a few ways to go about this:
 
 #### Upgrading nodes
-We will have `NodeKubeadmUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with the specified containers as `initContainers`. It will track their progress and bubble up any error/success to the `NodeKubeadmUpgrade` status. The status should also allow to track the progress of the different upgr ...
+We will have `NodeUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with `initContainers` with different commands depending on whether the node is a control plane node or a worker node. It will track their progress and bubble up any error/success to the `NodeUpgrade` status.
```
```suggestion
We will have `NodeUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with `initContainers` with different commands to the upgrader pod script depending on whether the node is a control plane node or a worker node. It will track their progress and bubble up any error/success to the `NodeUpgrade` status.
```

g-gaston commented:
Could we add in the PR description the justification for all the changes?

abhinavmpandey08 commented:
> I'm assuming we decided to make these changes to the design because we found reasons to do it and then we implemented that "new design". Could we add in the PR description the justification for all the changes?

I have added more details about the changes made in the design along with the justifications. Lmk if you have any more questions.

g-gaston commented:
> This was changed because the runtime extensions needed for this feature aren't defined in upstream CAPI yet but it will be added when in-place is officially supported upstream. Until then we would have to maintain a patch in our CAPI to add these runtime extensions which may or may not match what upstream CAPI would have.
Therefore the initial investment in adding runtime extensions doesn't seem justified considering we can achieve similar functionality by using annotations, which is a 2 way door decision. Once the runtime extension support for in-place upgrade is officially added in upstream CAPI, we can migrate to it without a lot of throwaway code.

Then I would suggest we also avoid patching the capi API. If we are concerned about implementing something that doesn't align with the final CAPI proposal (fair enough), the API is the most difficult thing to "backtrack", so we should make sure we don't change it until CAPI accepts the changes.

I would also suggest also that you somehow still present the original idea in the doc and just add more text explaining what the intermediate solution is that you are implementing and how it will be iterated towards the CAPI approved solution. That tell the reader what the final goal is and how our implementation will evolve over time.

> This was done because we can gather all the information that we need from the KCP/MD ref including the machines that need upgrade, the new Kubernetes version that the machines need to be upgraded to, and versions of other components that need to be upgraded.

This explains how the alternative solution works, not why it's better than the original proposal.

> This was done because most of the commands were generic between control plane node and worker nodes. The only exception was the kubeadm upgrade. So to account for this, the NodeUpgrade object will accept a NodeType field to signify whether the node to be upgraded should be a cp or a worker. Then the ControlPlaneUpgrade and the MachineDeploymentUpgrade controller can just pass the NodeType down without having to worry about what the exact implementation of what upgrading a control plane node or a worker node would entail, as those details are abstracted away.

Same here. Why is this better? I'm missing the reasoning behind the change.

Also I would love to see the tradeoffs. For example, you are leaking the "CP vs worker" concept to the node upgrader. You achieve a more compact API by paying the price on coupling. Now your node upgrader api and controller have multiple reasons to change and some of them are shared with the other controllers, which means some in situations you might have to change all the controllers at once.

rahulbabu95 commented:
/lgtm

abhinavmpandey08 commented on designs/in-place-upgrades-eksa.md:
```diff
@@ -98,15 +149,15 @@ The node upgrade process we need to perform, although different depending on the
 Draining and uncordoning the node could run in either the container or the host namespace. However, we will run it in the container namespace to be able to leverage the injected credentials for the `ServiceAccount`. This way we don't depend on having a kubeconfig in the host disk. This not only allows us to easily limit the rbac permissions that the `kubectl` command will use, but it's specially  ...
 
-In order to codify the logic of each step (the ones that require logic, like the kubeadm upgrade), we will build a single go binary with multiple commands (one per step). The `ControlPlaneKubeadmUpgrade` and `WorkersKubeadmUpgrade` will just reference these commands when building the init containers spec.
+In order to codify the logic of each step (the ones that require logic, like the kubeadm upgrade), we will build a single go binary with multiple commands (one per step). The `ControlPlaneUpgrade` and `MachineDeploymentUpgrade` will just reference these commands when building the init containers spec.
```
update: add a line about script temporarily instead of binary.
Update the last sentence to say `NodeUpgrade` instead of CPUpgrade and MDUpgrade

taneyland reviewed (commented):
Look great! Couple nits

abhinavmpandey08 commented:
/approve
